File: js/src/trees/isBalanced.js
class TreeNode {
	constructor(val = null, left = 0, right = 0) {
		this.val = val;
		this.left = left;
		this.right = right;
	}
}
const isBalanced = (root) => {
	const dfs = (current) => {
		if (!current) return [true, 0];

		const left = dfs(current.left);
		const right = dfs(current.right);

		const isBalanced = left[0] && right[0] && Math.abs(left[1] - right[1]) <= 1;
		return [isBalanced, 1 + Math.max(left[1], right[1])];
	};

	return dfs(root)[0];
};
const root = new TreeNode(
	1,
	new TreeNode(2, new TreeNode(4, new TreeNode(7))),
	new TreeNode(3),
);

console.log(isBalanced(root));

File: python/trees/isBalanced.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(current):
            if not current:
                return [True, 0]

            left, right = dfs(current.left), dfs(current.right)
            isBalanced = right[0] and left[0] and abs(left[1] - right[1]) <= 1
            return [isBalanced, 1 + max(left[1], right[1])]

        return dfs(root)[0]
